src/main/java/de/tum/in/www1/artemis/service/hestia/TestwiseCoverageService.java

```diff
@@ -0,0 +1,233 @@
+                // an ID and therefore cause an exception when the file report is saved to the database
+                var testwiseEntries = fileReport.getTestwiseCoverageEntries();
+
+                var optionalReportWithSameName = uniqueFileReports.stream().filter(report -> report.getFilePath().equals(fileReport.getFilePath())).findFirst();
+                CoverageFileReport savedFileReport;
+
+                if (optionalReportWithSameName.isEmpty()) {
+                    // Remove the testwise entries temporarily as they do not have an ID yet
+                    fileReport.setTestwiseCoverageEntries(Collections.emptySet());
+                    fileReport.setFullReport(savedFullReport);
+                    // Save the line count
+                    var lineCount = solutionLineCountByFilePath.get(fileReport.getFilePath());
+                    fileReport.setLineCount(lineCount);
+
+                    savedFileReport = coverageFileReportRepository.save(fileReport);
+                }
+                else {
+                    var reportWithSameName = optionalReportWithSameName.get();
+                    savedFileReport = coverageFileReportRepository.save(reportWithSameName);
+                }
+                uniqueFileReports.add(savedFileReport);
+
+                // Save all entries for the current file report to the database
+                testwiseEntries.forEach(entry -> {
+                    entry.setTestCase(testCase);
+                    entry.setFileReport(savedFileReport);
+                    testwiseCoverageReportEntryRepository.save(entry);
+                });
+            });
+        });
+
+        var updatedFullReport = coverageReportRepository.findByIdElseThrow(savedFullReport.getId());
+        // Calculate the unique line count for all file reports and save this value to the database
+        var coveredLinesCountByFilePath = calculateAndSaveUniqueLineCountsByFilePath(updatedFullReport);
+
+        // Calculate the aggregated covered line ratio over all files
+        double aggregatedCoveredLineRatio = calculateAggregatedLineCoverage(solutionLineCountByFilePath, coveredLinesCountByFilePath);
+        updatedFullReport.setCoveredLineRatio(aggregatedCoveredLineRatio);
+        coverageReportRepository.save(updatedFullReport);
+    }
+
+    /**
+     * Returns the line count by file name for all files in the solution repository for the last submission.
+     * @param programmingExercise the exercise from which the latest submission to the solution repository will be used
+     * @return line count by file name of files in the last submission's solution repository
+     */
+    private Map<String, Integer> getLineCountByFilePath(ProgrammingExercise programmingExercise) {
+        try {
+            var solutionParticipation = solutionProgrammingExerciseParticipationRepository.findByProgrammingExerciseIdElseThrow(programmingExercise.getId());
```

Comment:
You should not need to fetch this here. The Submission should already have the Participation fetched.